### PR TITLE
feat: add tool mode to components, Added current date tool to agent

### PR DESCRIPTION
**Pull request by edwinjosechittilappilly**
This pull request includes several updates to the `langflow` backend, focusing on enhancing the functionality of components and inputs by adding the `tool_mode` attribute, refining tool handling, and expanding test coverage.

Enhancements to components and inputs:
Added Current Date as tool to Agent.py

* [`src/backend/base/langflow/components/custom_component/custom_component.py`](diffhunk://#diff-6cdc1ec5642ada9f25a74e9e0f174b8b42d5e7cccda81055dcaa6dded67d432fL15-R15): Added `tool_mode` attribute to `MessageTextInput` in the `CustomComponent` class.
* [`src/backend/base/langflow/components/data/api_request.py`](diffhunk://#diff-b21ea5eba6b5613d5e6302b51e0f4940dc82f11bbafd75fe5fba59ac979edbf9R41-R42): Added `tool_mode` attribute to multiple inputs in the `APIRequestComponent` class and handled string type for `query_params` in `make_requests` method. [[1]](diffhunk://#diff-b21ea5eba6b5613d5e6302b51e0f4940dc82f11bbafd75fe5fba59ac979edbf9R41-R42) [[2]](diffhunk://#diff-b21ea5eba6b5613d5e6302b51e0f4940dc82f11bbafd75fe5fba59ac979edbf9R68) [[3]](diffhunk://#diff-b21ea5eba6b5613d5e6302b51e0f4940dc82f11bbafd75fe5fba59ac979edbf9R182-R185)
* [`src/backend/base/langflow/components/data/url.py`](diffhunk://#diff-be3b89252611954c8d88cb52a4f9a322d881951fe48d1bcd31d9b6af8838d794R24): Added `tool_mode` attribute to `URLComponent` class.
* [`src/backend/base/langflow/components/helpers/current_date.py`](diffhunk://#diff-6ddb06ce291adc301494245850a0c7c01286e80a85ed297a4b3165bd6344596eR59): Added `tool_mode` attribute to `CurrentDateComponent` class.
* [`src/backend/base/langflow/components/tools/calculator.py`](diffhunk://#diff-9e5e21486437724cadfa4fb3fcd27a293f1211870ea4c8e9ea48224be1233e1eR26): Added `tool_mode` attribute to `CalculatorToolComponent` class.
* [`src/backend/base/langflow/components/vectorstores/chroma.py`](diffhunk://#diff-7ea6e60d90c35d99062f4458018d4b9dd91b1db7a2f14cfbe21e8b8efa6e34a8R35): Added `tool_mode` attribute to `ChromaVectorStoreComponent` class.

Tool handling improvements:

* [`src/backend/base/langflow/base/tools/component_tool.py`](diffhunk://#diff-e6b607a8c5e404d2d308314cf694b33701ebcd21f8d564b6932c522d3a5e02cbL219-R219): Updated tool handling logic to check for `tool_name` and `tool_description` when there is only one tool.

Test coverage expansion:

* [`src/backend/tests/unit/custom/component/test_component_to_tool.py`](diffhunk://#diff-456ec68788a7d4c0a0cb802d0215df397e975adf88cd50ffa9932c6961512a94R3): Added a new test for the `CustomComponent` to ensure it correctly converts to a toolkit. [[1]](diffhunk://#diff-456ec68788a7d4c0a0cb802d0215df397e975adf88cd50ffa9932c6961512a94R3) [[2]](diffhunk://#diff-456ec68788a7d4c0a0cb802d0215df397e975adf88cd50ffa9932c6961512a94R33-R37)

These changes collectively enhance the functionality and robustness of the `langflow` backend components and inputs.
Waiting for #4402

### CODE REVIEW

**Comment by edwinjosechittilappilly** [comment]
@ogabrielluiz  Should I merge this into your branch (#4402 )  ? or should I merge into the main once #4402 is closed?

File: src/backend/base/langflow/components/agents/agent.py [hunk 1]
```diff
@@ -50,6 +58,8 @@ async def message_response(self) -> Message:
             msg = "No language model selected"
             raise ValueError(msg)
         self.chat_history = self.get_memory_data()
+        if self.add_current_date_tool:
+            self.tools.append(CurrentDateComponent().to_toolkit())
```
Comment (ogabrielluiz) [inline src/backend/base/langflow/components/agents/agent.py:62]: You should add a check because self.tools may not be a list.

**Review by ogabrielluiz — CHANGES_REQUESTED** [review]
Just this small fix and this should be good to go.

**Review by ogabrielluiz — APPROVED** [review]

**Comment by edwinjosechittilappilly** [comment]
Tested :
With the Fix: #4465 
New agents work as expected.

**Pull request merged**